From 1468daa9fcee2189c09431e0128a9d44a987ee04 Mon Sep 17 00:00:00 2001
From: jinghua <jinghua@marvell.com>
Date: Sun, 10 Jan 2016 05:32:50 +0800
Subject: [PATCH 0794/1240] a3700: enable multiple FDT suppor for a3700

- add fdt_config_id and board_id to two a3700 db dts files.
  * board_id = 0x10, for multi-fdt feature, a3700 db
    takes 0x10 board_id that starts from 0x10.
  * fdt_config_id is ID for each dts for a certain board.
    + armada-lp-db.dts: board_id = A3700_DB_CONFIG_ID_DEFAULT.
    + armada-lp-db-sgmii1.dts: board_id = A3700_DB_CONFIG_ID_SGMII1.

Change-Id: I50ebf2d4c712abc938914dcc85e32f824e88fe6a
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/26649
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
---
 arch/arm/dts/armada-lp-db-sgmii1.dts            |  2 +
 arch/arm/dts/armada-lp-db.dts                   |  3 ++
 arch/arm/include/asm/arch-armadalp/board-info.h | 52 +++++++++++++++++++++++++
 configs/mvebu_armadalp_defconfig                |  2 +-
 include/dt-bindings/multi-fdt/multi-fdt.h       | 41 +++++++++++++++++++
 5 files changed, 99 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/include/asm/arch-armadalp/board-info.h
 create mode 100644 include/dt-bindings/multi-fdt/multi-fdt.h

diff --git a/arch/arm/dts/armada-lp-db-sgmii1.dts b/arch/arm/dts/armada-lp-db-sgmii1.dts
index 115ea97..a3300c7 100644
--- a/arch/arm/dts/armada-lp-db-sgmii1.dts
+++ b/arch/arm/dts/armada-lp-db-sgmii1.dts
@@ -22,6 +22,8 @@
 / {
 	model = "DB-88F3720-DDR3-Modular-SGMII1";
 	compatible = "marvell,armada-lp-db0", "marvell,armada-lp";
+	fdt_config_id = <A3700_DB_CONFIG_ID_SGMII1>;
+	board_id = <A3700_DB_ID>;
 
 	soc {
 		internal-regs {
diff --git a/arch/arm/dts/armada-lp-db.dts b/arch/arm/dts/armada-lp-db.dts
index a4f405f..ab28d8e 100644
--- a/arch/arm/dts/armada-lp-db.dts
+++ b/arch/arm/dts/armada-lp-db.dts
@@ -2,6 +2,7 @@
 /dts-v1/;
 #include <dt-bindings/ddr/ddr_data.h>
 #include "armada-lp.dtsi"
+#include <dt-bindings/multi-fdt/multi-fdt.h>
 
 /* detail board setup:
  * Pcie Jumper slot	:	DB-88F3720-PCIe-mPCIe-Jumper
@@ -24,6 +25,8 @@
 / {
 	model = "DB-88F3720-DDR3-Modular";
 	compatible = "marvell,armada-lp-db0", "marvell,armada-lp";
+	fdt_config_id = <A3700_DB_CONFIG_ID_DEFAULT>;
+	board_id = <A3700_DB_ID>;
 
 	soc {
 		freq {
diff --git a/arch/arm/include/asm/arch-armadalp/board-info.h b/arch/arm/include/asm/arch-armadalp/board-info.h
new file mode 100644
index 0000000..0487267
--- /dev/null
+++ b/arch/arm/include/asm/arch-armadalp/board-info.h
@@ -0,0 +1,52 @@
+/*
+ * ***************************************************************************
+ * Copyright (C) 2015 Marvell International Ltd.
+ * ***************************************************************************
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation, either version 2 of the License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * ***************************************************************************
+ */
+
+/* Armada3700 Marvell boards */
+#include <dt-bindings/multi-fdt/multi-fdt.h>
+
+/*
+	ARMAMA3700_MARVELL_BOARD_ID_BASE = 0x10, for multi-fdt feature,
+	each Soc takes 0x10 boards that starts from 0x10, A8K has taken
+	0x10 to 0x1f, but since A8K and A3700 would not be compiled
+	together, to align with A8K, A3700 should also take 0x10 to 0x1f.
+	And as A3700 DB, it is the first board, should take board_id as 0x10.
+*/
+#define ARMAMA3700_MARVELL_BOARD_ID_BASE	 A3700_DB_ID
+#define ARMAMA3700_DB			(ARMAMA3700_MARVELL_BOARD_ID_BASE + 0)
+#define ARMAMA3700_MARVELL_MAX_BOARD_ID	(ARMAMA3700_MARVELL_BOARD_ID_BASE + 1)
+#define ARMAMA3700_MARVELL_BOARD_NUM	(ARMAMA3700_MARVELL_MAX_BOARD_ID - ARMAMA3700_MARVELL_BOARD_ID_BASE)
+
+#define MARVELL_BOARD_ID_BASE		ARMAMA3700_MARVELL_BOARD_ID_BASE
+#define MV_MAX_MARVELL_BOARD_ID		ARMAMA3700_MARVELL_MAX_BOARD_ID
+#define MV_MARVELL_BOARD_NUM		ARMAMA3700_MARVELL_BOARD_NUM
+#define MV_DEFAULT_BOARD_ID		ARMAMA3700_DB
+
+#define BOARD_DEV_TWSI_INIT_EEPROM 0x57
+#define MULTI_FDT_EEPROM_ADDR_LEN 2
+
+/*
+	fdt_config_id is ID for each dts for a certain board.
+	It starts from 1.
+	+ arch/arm/dts/armada-lp-db.dts, board_id = A3700_DB_CONFIG_ID_USB3..
+	+ arch/arm/dts/armada-lp-db-sgmii1.dts, board_id = A3700_DB_CONFIG_ID_SGMII1.
+*/
+#define DEFAULT_FDT_CONFIG_ID	A3700_DB_CONFIG_ID_USB3
+
+#define DEFAULT_FDT_PER_BOARD { \
+	DEFAULT_FDT_CONFIG_ID,	/* default fdt_config_id for armada-3700-db board */	\
+}
diff --git a/configs/mvebu_armadalp_defconfig b/configs/mvebu_armadalp_defconfig
index c00060a..2aa9410 100644
--- a/configs/mvebu_armadalp_defconfig
+++ b/configs/mvebu_armadalp_defconfig
@@ -10,7 +10,6 @@ CONFIG_XENON_MMC=y
 CONFIG_MVEBU_MPP_BUS=y
 CONFIG_MVEBU_PINCTL=y
 +S:CONFIG_DEVEL_BOARD=y
-+S:# CONFIG_MULTI_DT_FILE is not set
 CONFIG_CMD_BDI=y
 CONFIG_CMD_BOOTD=y
 CONFIG_CMD_RUN=y
@@ -22,6 +21,7 @@ CONFIG_CMD_NET=y
 CONFIG_CMD_PING=y
 CONFIG_CMD_MII=y
 CONFIG_CMD_MVEBU_MPP=y
+CONFIG_CMD_MVEBU_FDT_CONFIG=y
 CONFIG_CMD_MVEBU_BUBT=y
 +S:CONFIG_OF_CONTROL=y
 +S:CONFIG_OF_EMBED=y
diff --git a/include/dt-bindings/multi-fdt/multi-fdt.h b/include/dt-bindings/multi-fdt/multi-fdt.h
new file mode 100644
index 0000000..74aab4d
--- /dev/null
+++ b/include/dt-bindings/multi-fdt/multi-fdt.h
@@ -0,0 +1,41 @@
+/*
+ * ***************************************************************************
+ * Copyright (C) 2015 Marvell International Ltd.
+ * ***************************************************************************
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation, either version 2 of the License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * ***************************************************************************
+ */
+
+#ifndef _MULTI_FDT_H_
+#define _MULTI_FDT_H_
+
+#define A3700_DB_ID			0x10
+/*
+ * There are two main presets for Armada3700 DB, between
+ * these two presets, most of the configurations are the same:
+   - CPU 600MHz and DDR 600MHz
+   - SATA
+   - neta0 - RGMII
+   - PCIe(COMPHY-0)
+ *
+ * The only difference is COMPHY-1
+ * preset 1: COMPHY-1 is working as USB3.
+ * preset 2. COMPHY-1 is working as SGMII-1
+ */
+#define A3700_DB_CONFIG_ID_USB3		1
+#define A3700_DB_CONFIG_ID_SGMII1	2
+
+#define A3700_DB_CONFIG_ID_DEFAULT	A3700_DB_CONFIG_ID_USB3
+
+#endif /* _MULTI_FDT_H_ */
+
-- 
1.9.1

